File: src/events/ready.ts
import cron from "cron";
import {
  ActivityType,
  ChannelType,
  type Client,
  EmbedBuilder,
  PermissionsBitField,
} from "discord.js";
import config from "../config/config";
import { botHasRecentMessages, getServerData } from "../utils/helpers";
import { existsRoleMenu, sendRoleMenu } from "../utils/menu";
import { sendReactionRoleMenu } from "../utils/reactionMenu";
import { sendYoutubeNotification } from "../utils/youtube";
import { sync } from "./sync";

export const onReady = async (client: Client) => {
  console.log("[Ready] Bot initialization starting...");
  console.log(
    `[Ready] Logged in as ${client.user?.tag} (ID: ${client.user?.id})`,
  );
  console.log(`[Ready] Connected to ${client.guilds.cache.size} guild(s)`);

  const cronLog = (job: string, message: string) =>
    console.log(`[Cron][${job}] ${message}`);

  async function runSync() {
    console.log("Running sync...");
    try {
      if (client.isReady()) {
        await sync(client);
      }
    } finally {
      setTimeout(runSync, 30_000);
    }
  }

  async function sendReminder() {
    cronLog("SendReminder", "Fetching channel");
    const channel = await client.channels.fetch(config.generalId);
    if (channel && channel.type === ChannelType.GuildText) {
      cronLog("SendReminder", `Sending reminder to channel ${channel.id}`);
      await channel.send(config.generalMessage);
    } else {
      console.error(
        `Couldn't find general channel by ID: ${config.generalId} ${channel}`,
      );
    }
    cronLog("SendReminder", "Finished");
  }

  async function sendNotification() {
    cronLog("SendNotification", "Fetching YouTube channel");
    const youtubeChannel = await client.channels.fetch(config.youtubeId);
    if (!youtubeChannel) {
      console.error(
        `Couldn't find youtube videos channel by ID: ${config.youtubeId} ${youtubeChannel}`,
      );
      return;
    }

    if (youtubeChannel.type !== ChannelType.GuildAnnouncement) {
      console.error(
        `Youtube videos channel (${config.youtubeId} ${youtubeChannel}) isn't an announcement channel`,
      );
      return;
    }

    cronLog("SendNotification", "Sending YouTube notification");
    await sendYoutubeNotification(
      youtubeChannel,
      config.youtubeQueries,
      config.youtubeIgnoreWords,
      config.youtubeWhitelistedIds,
    );
    cronLog("SendNotification", "Finished");
  }

  async function sendRoleMenuMsg() {
    cronLog("SendRoleMenuMsg", "Fetching role menu channel");
    const roleChannel = await client.channels.fetch(config.roleMenuId);
    if (!roleChannel) {
      console.error(
        `Couldn't find role menu channel by ID: ${config.roleMenuId} ${roleChannel}`,
      );
      return;
    }

    if (roleChannel.type !== ChannelType.GuildText) {
      console.error(
        `Role menu channel (${config.roleMenuId} ${roleChannel}) isn't a text channel`,
      );
      return;
    }

    cronLog("SendRoleMenuMsg", "Checking if role menu already exists");
    const existsMenu = await existsRoleMenu(roleChannel);
    if (existsMenu) {
      cronLog("SendRoleMenuMsg", "Role menu already exists, skipping");
      return;
    }

    cronLog("SendRoleMenuMsg", "Sending role menu");
    await sendRoleMenu(roleChannel);
    cronLog("SendRoleMenuMsg", "Finished");
  }

  async function cleanRoleMenuRoles() {
    cronLog("CleanRoleMenuRoles", "Fetching guild");
    const guild = await client.guilds.fetch(config.guildId);
    if (!guild) {
      cronLog("CleanRoleMenuRoles", "Guild not found, skipping");
      return;
    }

    try {
      await guild.members.fetch();
    } catch (error) {
      console.error(
        "[Cron][CleanRoleMenuRoles] Failed to fetch guild members:",
        error,
      );
      return;
    }

    const memberIdsWithMenuRoles = new Set<string>();

    for (const roleId of config.roleMenuRoleIds) {
      try {
        const role =
          guild.roles.cache.get(roleId) ?? (await guild.roles.fetch(roleId));
        if (!role) {
          cronLog("CleanRoleMenuRoles", `Role ${roleId} not found, skipping`);
          continue;
        }

        role.members.forEach((member) => {
          memberIdsWithMenuRoles.add(member.id);
        });
      } catch (error) {
        console.error(
          `[Cron][CleanRoleMenuRoles] Failed to inspect role ${roleId}:`,
          error,
        );
      }
    }

    let removedCount = 0;
    for (const memberId of memberIdsWithMenuRoles) {
      const member = guild.members.cache.get(memberId);
      if (!member) continue;

      const hasAccess =
        member.roles.cache.has(config.roleMenuRequiredRoleId) ||
        member.permissions.has(PermissionsBitField.Flags.Administrator);

      if (hasAccess) continue;

      try {
        await member.roles.remove(config.roleMenuRoleIds);
        removedCount += 1;
      } catch (error) {
        console.error(
          `[Cron][CleanRoleMenuRoles] Failed to remove role menu roles from ${member.id}:`,
          error,
        );
      }
    }

    cronLog(
      "CleanRoleMenuRoles",
      `Removed role menu roles from ${removedCount} member(s)`,
    );
  }

  async function sendReactionRoleMenus() {
    cronLog("SendReactionRoleMenus", "Fetching reaction role channel");
    const reactionRoleChannel = await client.channels.fetch(
      config.reactionRoleMenuId,
    );
    if (!reactionRoleChannel) {
      console.error(
        `Couldn't find reaction role channel by ID: ${config.reactionRoleMenuId} ${reactionRoleChannel}`,
      );
      return;
    }

    if (reactionRoleChannel.type !== ChannelType.GuildText) {
      console.error(
        `Reaction role channel (${config.reactionRoleMenuId} ${reactionRoleChannel}) isn't a text channel`,
      );
      return;
    }

    cronLog("SendReactionRoleMenus", "Building embed messages");
    const embeds = [
      new EmbedBuilder()
        .setAuthor({
          name: "6b6t.org",
          iconURL: "https://www.6b6t.org/logo.png",
        })
        .setDescription(
          `
Select your language.
      `,
        )
        .setColor("#07CFFA"),
      new EmbedBuilder()
        .setAuthor({
          name: "6b6t.org",
          iconURL: "https://www.6b6t.org/logo.png",
        })
        .setImage("https://www.6b6t.org/media/language-and-roles.gif")
        .setDescription(
          `
Select your notifications.

✨ - General changes to 6b6t
⚔️ - Crystal PvP, anticheat changes, PvP events and more
🌩️ - Server going offline, online or restarting
🎉 - Events and competitions in Discord and Minecraft
🏄 - Help us test new features
🎥 - Receive social media notifications
      `,
        )
        .setColor("#FFF11A"),
      new EmbedBuilder()
        .setAuthor({
          name: "6b6t.org",
          iconURL: "https://www.6b6t.org/logo.png",
        })
        .setDescription(
          `
🎮 - Get notifications about Hytale.
      `,
        )
        .setColor("#82c0ef"),
    ];

    cronLog("SendReactionRoleMenus", "Checking for existing bot messages");
    const messageCount = await botHasRecentMessages(
      reactionRoleChannel,
      client,
    );
    if (messageCount < embeds.length) {
      cronLog(
        "SendReactionRoleMenus",
        `Found ${messageCount} existing messages (need ${embeds.length}), skipping`,
      );
      return;
    }

    cronLog(
      "SendReactionRoleMenus",
      `Sending ${embeds.length} reaction role menu(s)`,
    );
    for (const embed of embeds) {
      await sendReactionRoleMenu(
        reactionRoleChannel,
        config.languageMenuRoleIds,
        embed,
      );
    }
    cronLog("SendReactionRoleMenus", "Finished");
  }

  async function updateStatus() {
    if (!client.user) {
      cronLog("UpdateStatus", "Client user unavailable, skipping");
      return;
    }
    cronLog("UpdateStatus", "Fetching server data");
    const data = await getServerData();
    if (!data) {
      cronLog("UpdateStatus", "No data received from server");
      return;
    }
    cronLog(
      "UpdateStatus",
      `Setting activity with ${data.playerCount} players online`,
    );
    client.user.setActivity(
      `IP: ${config.statusHost} - Join ${data.playerCount} other players online!`,
      { type: ActivityType.Playing },
    );
    cronLog("UpdateStatus", "Finished");
  }

  console.log("[Ready] Running initial status update...");
  await updateStatus();
  console.log("[Ready] Initial status update complete");

  console.log("[Ready] Scheduling cron jobs...");
  new cron.CronJob("*/5 * * * *", updateStatus, null, true, "Europe/Berlin");
  console.log("[Ready] Scheduled UpdateStatus cron (every 5 minutes)");

  new cron.CronJob(
    "*/5 * * * *",
    cleanRoleMenuRoles,
    null,
    true,
    "Europe/Berlin",
  );
  console.log("[Ready] Scheduled CleanRoleMenuRoles cron (every 5 minutes)");

  new cron.CronJob("0 10 * * *", sendReminder, null, true, "Europe/Berlin");
  console.log("[Ready] Scheduled SendReminder cron (daily at 10:00)");

  new cron.CronJob("0 18 * * *", sendReminder, null, true, "Europe/Berlin");
  console.log("[Ready] Scheduled SendReminder cron (daily at 18:00)");

  /*
  Free trial is 100 checks a day, which is 1 check every 15 minutes, 20 minutes to be safe
  This would ignore any video sent before those 20 minutes, but it can't be fixed without paying or using IFTTT
  */

  new cron.CronJob(
    "*/20 * * * *",
    sendNotification,
    null,
    true,
    "Europe/Berlin",
  );
  console.log("[Ready] Scheduled SendNotification cron (every 20 minutes)");

  console.log("[Ready] Starting initial background tasks...");
  console.log("[Ready] Starting SendNotification...");
  void sendNotification();
  console.log("[Ready] Starting SendRoleMenuMsg...");
  void sendRoleMenuMsg();
  console.log("[Ready] Starting CleanRoleMenuRoles...");
  void cleanRoleMenuRoles();
  console.log("[Ready] Starting SendReactionRoleMenus...");
  void sendReactionRoleMenus();
  console.log("[Ready] Starting RunSync...");
  void runSync();

  console.log("[Ready] Bot initialization complete - all tasks started");
};

File: src/events/ready.rs
use std::{collections::HashSet, future::Future, time::Duration};

use anyhow::Result;
use chrono::{Datelike as _, Timelike as _};
use chrono_tz::Europe::Berlin;
use futures::TryStreamExt as _;
use poise::serenity_prelude as serenity;

use crate::{config, state::AppState};

use super::role_sync;

pub async fn handle(
    ctx: &serenity::Context,
    data: &AppState,
    ready: &serenity::Ready,
) -> Result<()> {
    tracing::info!(user = %ready.user.tag(), guilds = ready.guilds.len(), "Discord bot connected");
    let mut started = data.ready_started.lock().await;
    if *started {
        return Ok(());
    }
    *started = true;
    drop(started);

    if let Some(telegram) = &data.telegram {
        let ctx = ctx.clone();
        let telegram = telegram.clone();
        spawn_startup_retry("Telegram initialization", move || {
            let ctx = ctx.clone();
            let telegram = telegram.clone();
            async move { telegram.ready(&ctx).await }
        });
    }
    let role_menu_ctx = ctx.clone();
    spawn_startup_retry("role menu initialization", move || {
        let ctx = role_menu_ctx.clone();
        async move { ensure_role_menu(&ctx).await }
    });
    let reaction_menu_ctx = ctx.clone();
    spawn_startup_retry("reaction menu initialization", move || {
        let ctx = reaction_menu_ctx.clone();
        async move { ensure_reaction_menus(&ctx).await }
    });

    spawn_interval(
        ctx.clone(),
        data.clone(),
        Duration::from_mins(5),
        |ctx, data| Box::pin(update_status(ctx, data)),
    );
    spawn_interval(
        ctx.clone(),
        data.clone(),
        Duration::from_mins(5),
        |ctx, data| Box::pin(clean_role_menu_roles(ctx, data)),
    );
    spawn_interval(
        ctx.clone(),
        data.clone(),
        Duration::from_mins(20),
        |ctx, data| Box::pin(youtube_notification(ctx, data)),
    );
    spawn_interval(
        ctx.clone(),
        data.clone(),
        Duration::from_secs(30),
        |ctx, data| Box::pin(role_sync::run(ctx, data)),
    );
    spawn_interval(
        ctx.clone(),
        data.clone(),
        Duration::from_mins(10),
        |_ctx, data| Box::pin(clean_pending_approvals(data)),
    );
    if data.anarchy.is_some() {
        spawn_interval(
            ctx.clone(),
            data.clone(),
            Duration::from_hours(1),
            |ctx, data| Box::pin(anarchy_analytics(ctx, data)),
        );
    }
    if data.community_event.is_some() {
        spawn_interval(
            ctx.clone(),
            data.clone(),
            Duration::from_secs(5),
            |ctx, data| Box::pin(community_event_announcements(ctx, data)),
        );
    }
    spawn_reminders(ctx.clone());
    Ok(())
}

fn spawn_startup_retry<F, Fut>(label: &'static str, task: F)
where
    F: Fn() -> Fut + Send + Sync + 'static,
    Fut: Future<Output = Result<()>> + Send + 'static,
{
    tokio::spawn(async move {
        loop {
            match task().await {
                Ok(()) => return,
                Err(error) => {
                    tracing::error!(%error, task = label, "startup task failed; retrying");
                    tokio::time::sleep(Duration::from_mins(1)).await;
                }
            }
        }
    });
}

fn spawn_interval<F>(ctx: serenity::Context, data: AppState, interval: Duration, task: F)
where
    F: for<'a> Fn(&'a serenity::Context, &'a AppState) -> futures::future::BoxFuture<'a, ()>
        + Send
        + Sync
        + 'static,
{
    tokio::spawn(async move {
        task(&ctx, &data).await;
        let mut ticker = tokio::time::interval(interval);
        ticker.tick().await;
        loop {
            ticker.tick().await;
            task(&ctx, &data).await;
        }
    });
}

fn spawn_reminders(ctx: serenity::Context) {
    tokio::spawn(async move {
        let mut sent = HashSet::new();
        let mut ticker = tokio::time::interval(Duration::from_secs(30));
        loop {
            ticker.tick().await;
            let now = chrono::Utc::now().with_timezone(&Berlin);
            if now.minute() != 0 || !matches!(now.hour(), 10 | 18) {
                continue;
            }
            let key = (now.year(), now.ordinal(), now.hour());
            if !sent.insert(key) {
                continue;
            }
            sent.retain(|(year, ordinal, _)| *year == now.year() && *ordinal + 2 >= now.ordinal());
            if let Err(error) = config::GENERAL_ID.say(&ctx, config::GENERAL_MESSAGE).await {
                tracing::error!(%error, "failed to send general rank reminder");
            }
        }
    });
}

async fn update_status(ctx: &serenity::Context, data: &AppState) {
    match data.server.server_data().await {
        Ok(server) => ctx.set_activity(Some(serenity::ActivityData::playing(format!(
            "IP: play.6b6t.org - Join {} other players online!",
            server.player_count
        )))),
        Err(error) => tracing::error!(%error, "failed to update Discord presence"),
    }
}

async fn youtube_notification(ctx: &serenity::Context, data: &AppState) {
    if let Err(error) = data.youtube.notify(ctx, config::YOUTUBE_ID).await {
        tracing::error!(%error, "YouTube notification check failed");
    }
}

async fn clean_pending_approvals(data: &AppState) {
    let removed = data.pending.cleanup_expired().await;
    if removed > 0 {
        tracing::info!(removed, "expired pending approval requests");
    }
}

async fn anarchy_analytics(ctx: &serenity::Context, data: &AppState) {
    let Some(anarchy) = &data.anarchy else {
        return;
    };
    let (online_users, online_players) = tokio::join!(
        data.server.anarchymod_player_count(),
        data.server.player_count(),
    );
    let online_users = match online_users {
        Ok(count) => Some(count),
        Err(error) => {
            tracing::warn!(%error, "failed to fetch the online AnarchyMod player count; the count is shown as unavailable");
            None
        }
    };
    let online_players = match online_players {
        Ok(count) => Some(count),
        Err(error) => {
            tracing::warn!(%error, "failed to fetch the player count for anarchy analytics; the online percentage is omitted");
            None
        }
    };
    if let Err(error) = anarchy.report(ctx, online_users, online_players).await {
        tracing::error!(%error, "anarchy mod analytics report failed");
    }
}

async fn community_event_announcements(ctx: &serenity::Context, data: &AppState) {
    let Some(service) = &data.community_event else {
        return;
    };
    if let Err(error) = service.poll(ctx).await {
        tracing::error!(%error, "community-event Discord announcement check failed");
    }
}

async fn clean_role_menu_roles(ctx: &serenity::Context, _data: &AppState) {
    let members = match config::GUILD_ID
        .members_iter(&ctx.http)
        .try_collect::<Vec<_>>()
        .await
    {
        Ok(members) => members,
        Err(error) => {
            tracing::error!(%error, "failed to load members for color role cleanup");
            return;
        }
    };
    for member in members {
        let has_menu_role = member
            .roles
            .iter()
            .any(|role| config::ROLE_MENU_ROLE_IDS.contains(role));
        let is_administrator = ctx.cache.guild(config::GUILD_ID).is_some_and(|guild| {
            guild.owner_id == member.user.id
                || member.roles.iter().any(|role_id| {
                    guild
                        .roles
                        .get(role_id)
                        .is_some_and(|role| role.permissions.administrator())
                })
        });
        let has_access =
            member.roles.contains(&config::ROLE_MENU_REQUIRED_ROLE_ID) || is_administrator;
        if has_menu_role && !has_access {
            for role in config::ROLE_MENU_ROLE_IDS {
                if member.roles.contains(role)
                    && let Err(error) = member.remove_role(ctx, *role).await
                {
                    tracing::error!(%error, member_id = %member.user.id, role_id = %role, "failed to remove inaccessible color role");
                }
            }
        }
    }
}

async fn ensure_role_menu(ctx: &serenity::Context) -> Result<()> {
    let messages = config::ROLE_MENU_ID
        .messages(ctx, serenity::GetMessages::new().limit(10))
        .await?;
    let exists = messages.iter().flat_map(|message| &message.components).flat_map(|row| &row.components).any(|component| matches!(component, serenity::ActionRowComponent::SelectMenu(menu) if menu.custom_id.as_deref() == Some("legend_role_menu")));
    if exists {
        return Ok(());
    }
    let roles = config::GUILD_ID.roles(ctx).await?;
    let mut options = vec![serenity::CreateSelectMenuOption::new("Clear", "clear_top")];
    options.extend(
        config::ROLE_MENU_ROLE_IDS
            .iter()
            .filter_map(|id| roles.get(id))
            .map(|role| serenity::CreateSelectMenuOption::new(&role.name, role.id.to_string())),
    );
    options.push(serenity::CreateSelectMenuOption::new(
        "Clear",
        "clear_bottom",
    ));
    let menu = serenity::CreateSelectMenu::new(
        "legend_role_menu",
        serenity::CreateSelectMenuKind::String { options },
    )
    .placeholder("Select a color");
    let embed = serenity::CreateEmbed::new()
        .title("Legend Color Roles")
        .description("Change your color in the Discord by picking one of the colors.")
        .image("https://www.6b6t.org/media/legend-color.gif")
        .thumbnail("https://www.6b6t.org/logo.png")
        .colour(0x00FF_F11A);
    config::ROLE_MENU_ID
        .send_message(
            ctx,
            serenity::CreateMessage::new()
                .embed(embed)
                .components(vec![serenity::CreateActionRow::SelectMenu(menu)]),
        )
        .await?;
    Ok(())
}

async fn ensure_reaction_menus(ctx: &serenity::Context) -> Result<()> {
    let messages = config::REACTION_ROLE_MENU_ID
        .messages(ctx, serenity::GetMessages::new().limit(100))
        .await?;
    let existing = messages
        .iter()
        .filter(|message| message.author.id == ctx.cache.current_user().id)
        .count();
    if existing >= 3 {
        return Ok(());
    }
    let menus = [
        (
            "Select your language.",
            0x0007_CFFA,
            &config::REACTION_ROLES[6..12],
            None,
        ),
        (
            "Select your notifications.\n\n✨ - General changes to 6b6t\n⚔️ - Crystal PvP, anticheat changes, PvP events and more\n🌩️ - Server going offline, online or restarting\n🎉 - Events and competitions in Discord and Minecraft\n🏄 - Help us test new features\n🎥 - Receive social media notifications",
            0x00FF_F11A,
            &config::REACTION_ROLES[..6],
            Some("https://www.6b6t.org/media/language-and-roles.gif"),
        ),
        (
            "🎮 - Get notifications about Hytale.",
            0x0082_C0EF,
            &config::REACTION_ROLES[12..],
            None,
        ),
    ];
    for (description, colour, roles, image) in menus {
        let mut embed = serenity::CreateEmbed::new()
            .author(
                serenity::CreateEmbedAuthor::new("6b6t.org")
                    .icon_url("https://www.6b6t.org/logo.png"),
            )
            .description(description)
            .colour(colour);
        if let Some(image) = image {
            embed = embed.image(image);
        }
        let message = config::REACTION_ROLE_MENU_ID
            .send_message(ctx, serenity::CreateMessage::new().embed(embed))
            .await?;
        for (emoji, _) in roles {
            message
                .react(ctx, serenity::ReactionType::Unicode((*emoji).to_owned()))
                .await?;
        }
    }
    Ok(())
}
